blog/_posts/06-03-blogpost.md

```diff
@@ -0,0 +1,164 @@
+
+There's also the secondary small BitVector optimization which will involve adding an extra byte of control bits before the size and cannibalizing the eight bits of size for extra space to grant BitVector the same optimization for 128 bits and fewer.  
+
+## DataMap
+
+A DataMap is two other classes in a trench coat: a pointer to a DataLayout and a MemoryImage.  Each MemoryImage belongs to one and only one DataMap, but DataLayouts can be linked to several DataMaps.  The DataLayout contains information about the MemoryImage, such as the names (strings), indices (size_ts), descriptions and notes (strings), and such things as constructors and destructors for more complex objects.  The MemoryImage is a vector of bytes with each object taking up a number of them in contiguous memory.  Much of the MemoryImage's deeper workings rely on reinterpret casts and memory tricks (which I won't go into), but the main gist of it is that it contains a region of memory which holds a series of items of various types.  On the surface, however, you don't ever need to interact with DataLayout or MemoryImage.  Add variables using AddVar\<T\>(string name, T value, string description, string notes), which returns a size_t which is the variable's index in the Vector of bytes.  Now, you can access that variable with Get\<T\>(string name) or Get\<T\>(size_t id), but the id access is significantly faster.  You can also set the variable again with Set\<T\>(string name, T value) or Set\<T\>(size_t id, T value).  There are other methods involving comparing layouts and getting types and seeing if ids and names exist, but those are far less important for the day-to-day use of DataMap.
+
+Compared to a regular Map, which is a hash table, the DataMap is far more efficient if fewer additions and deletions are being made compared to sets and gets.  In fact, DataMaps don't even allow for deletions at all.  As such, it is more optimized for scientific computing and the large number of accesses which are commonplace in MABE.
+
+However, as I said before, this rigidity poses a bit of a flaw.  What if you add a variable of type Foo.  This Foo is added to the DataMap and now exists in DataMap's memory.  But what happens if you want to add another Foo?  Well, you have to take that other Foo and replace the first Foo in DataMap's memory.  But what if this Foo is someplace else, and can change?  Well, you're out of luck.  But if we take the memory DataMap gives for that Foo, and we place a pointer there instead (with a byte of control bits, of course), we can handle this sort of situation.
```

Comment:
This section is well explained/laid out for someone who doesn't have prior exposure to DataMaps. Again, I really like the use of motivational questions here to summarize the above points you make